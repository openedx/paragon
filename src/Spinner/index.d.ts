import React from 'react';
import { SpinnerProps as BaseSpinnerProps } from 'react-bootstrap/Spinner';

export interface SpinnerProps extends BaseSpinnerProps {
    className?: string;
    screenReaderText?: React.ReactNode;
}

declare const Spinner: React.ForwardRefExoticComponent<SpinnerProps>;

export default Spinner;
