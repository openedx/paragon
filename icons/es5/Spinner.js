function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import * as React from "react";

function SvgSpinner(props) {
  return /*#__PURE__*/React.createElement("svg", _extends({
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), /*#__PURE__*/React.createElement("path", {
    d: "M19 8l-4 4h3c0 3.31-2.69 6-6 6a5.87 5.87 0 01-2.8-.7l-1.46 1.46A7.93 7.93 0 0012 20c4.42 0 8-3.58 8-8h3l-4-4zM6 12c0-3.31 2.69-6 6-6 1.01 0 1.97.25 2.8.7l1.46-1.46A7.93 7.93 0 0012 4c-4.42 0-8 3.58-8 8H1l4 4 4-4H6z",
    fill: "currentColor"
  }));
}

export default SvgSpinner;