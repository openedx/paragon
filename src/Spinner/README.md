---
title: 'Spinner'
type: 'component'
status: 'Stable'
components:
- Spinner
categories:
- Status & metadata
- Choreography
designStatus: 'Done'
devStatus: 'Done'
notes: |

---

A spinning animation that indicates loading.

<p>
  This is a pass through component from React-Bootstrap.<br/>
  <a href="https://react-bootstrap.github.io/components/spinners" target="_blank" rel="noopener noreferrer">
    See React-Bootstrap for documentation.
  </a>
</p>

## Basic Usage

```jsx live
<>
  <Spinner animation="border" className="mie-3" screenReaderText="loading" />
  <Spinner animation="grow" className="mie-3" screenReaderText="loading" />
</>
```
## Color Variants

```jsx live
() => {
  const variants = ['primary', 'secondary', 'success', 'danger', 'warning', 'info', 'light', 'dark'];
  return variants.map((variant) => (
    <Spinner animation="border" variant={variant} className="mr-3" screenReaderText="loading" />
  ));
}
```
