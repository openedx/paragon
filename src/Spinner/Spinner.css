/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:20 GMT
 */

:root {
  --pgn-size-spinner-sm-border-width: .2em;
  --pgn-size-spinner-sm-width: 1rem;
  --pgn-size-spinner-base-border-width: .25em;
  --pgn-size-spinner-base-width: 2rem;
  --pgn-size-spinner-sm-height: var(--pgn-size-spinner-sm-width);
  --pgn-size-spinner-base-height: var(--pgn-size-spinner-base-width);
}
