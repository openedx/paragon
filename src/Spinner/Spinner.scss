//
// Rotating border
//

@keyframes spinner-border {
  to { transform: rotate(360deg); }
}

.spinner-border {
  display: inline-block;
  width: var(--pgn-size-spinner-base-width);
  height: var(--pgn-size-spinner-base-height);
  vertical-align: $spinner-vertical-align;
  border: var(--pgn-size-spinner-base-border-width) solid currentcolor;
  border-right-color: transparent;
  // stylelint-disable-next-line property-disallowed-list
  border-radius: 50%;
  animation: .75s linear infinite spinner-border;
}

.spinner-border-sm {
  width: var(--pgn-size-spinner-sm-width);
  height: var(--pgn-size-spinner-sm-height);
  border-width: var(--pgn-size-spinner-sm-border-width);
}

//
// Growing circle
//

@keyframes spinner-grow {
  0% {
    transform: scale(0);
  }
  50% {
    opacity: 1;
    transform: none;
  }
}

.spinner-grow {
  display: inline-block;
  width: var(--pgn-size-spinner-base-width);
  height: var(--pgn-size-spinner-base-height);
  vertical-align: $spinner-vertical-align;
  background-color: currentcolor;
  // stylelint-disable-next-line property-disallowed-list
  border-radius: 50%;
  opacity: 0;
  animation: .75s linear infinite spinner-grow;
}

.spinner-grow-sm {
  width: var(--pgn-size-spinner-sm-width);
  height: var(--pgn-size-spinner-sm-height);
}

@if $enable-prefers-reduced-motion-media-query {
  @media (prefers-reduced-motion: reduce) {
    .spinner-border,
    .spinner-grow {
      animation-duration: 1.5s;
    }
  }
}

