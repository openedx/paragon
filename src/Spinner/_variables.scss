// Spinners

$spinner-width:         2rem !default;
$spinner-height:        $spinner-width !default;
$spinner-border-width:  .25em !default;

$spinner-width-sm:        1rem !default;
$spinner-height-sm:       $spinner-width-sm !default;
$spinner-border-width-sm: .2em !default;
