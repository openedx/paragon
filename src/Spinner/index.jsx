import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import BaseSpinner from 'react-bootstrap/Spinner';

const Spinner = React.forwardRef(({
  className,
  screenReaderText,
  ...attrs
}, ref) => {
  const spinnerProps = {
    ...attrs,
    className: classNames('pgn__spinner', className),
    role: screenReaderText ? 'status' : undefined,
  };
  return (
    <BaseSpinner {...spinnerProps} ref={ref}>
      {screenReaderText && <span className="sr-only">{screenReaderText}</span>}
    </BaseSpinner>
  );
});

Spinner.propTypes = {
  /** Specifies the class name for the component. */
  className: PropTypes.string,
  /** Specifies the screen reader content for a11y. */
  screenReaderText: PropTypes.node,
};

Spinner.defaultProps = {
  className: undefined,
  screenReaderText: undefined,
};

export default Spinner;
