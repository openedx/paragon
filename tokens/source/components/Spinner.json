{
  "spinner": {
    "width": { "value": "2rem", "themeable": true, "source": "$spinner-width" },
    "height": { "value": "{spinner.width.value}", "themeable": true, "source": "$spinner-height" },
    "border-width": { "value": ".25em", "themeable": true, "source": "$spinner-border-width" },
    "sm": {
      "width": { "value": "1rem", "themeable": true, "source": "$spinner-width-sm" },
      "height": { "value": "{spinner.sm.width}", "themeable": true, "source": "$spinner-height-sm" },
      "border-width": { "value": ".2em", "themeable": true, "source": "$spinner-border-width-sm" }
    }
  }
}
