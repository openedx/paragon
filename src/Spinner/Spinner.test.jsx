import React from 'react';
import { shallow } from 'enzyme';

import Spinner from './index';

describe('Spinner', () => {
  it('should render a spinner', () => {
    const wrapper = shallow(<Spinner animation="border" />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should render a spinner with screen reader text', () => {
    const wrapper = shallow(<Spinner animation="border" screenReaderText="loading" />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
