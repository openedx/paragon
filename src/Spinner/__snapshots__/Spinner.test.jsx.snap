// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner should render a spinner 1`] = `"<div class=\\"pgn__spinner spinner-border\\"></div>"`;

exports[`Spinner should render a spinner with screen reader text 1`] = `"<div role=\\"status\\" class=\\"pgn__spinner spinner-border\\"><span class=\\"sr-only\\">loading</span></div>"`;
